packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1227,6 +1249,95 @@ class RenderEditable extends RenderBox {
     }
   }
 
+  /// Sets the screen position of the floating cursor and the text position
+  /// closest to the cursor.
+  void setFloatingCursor(TextCursorState state, Offset boundedOffset, TextPosition lastTextPosition) {
+    if (state == TextCursorState.Start) {
+      _relativeOrigin = const Offset(0, 0);
+      _previousOffset = null;
+      _resetOriginOnBottom = false;
+      _resetOriginOnTop = false;
+      _resetOriginOnRight = false;
+      _resetOriginOnBottom = false;
+    }
+    _floatingCursorOn = state != TextCursorState.End;
+    if(_floatingCursorOn) {
+      markNeedsPaint();
+      _floatingCursorOffset = boundedOffset;
+      _floatingCursorTextPosition = lastTextPosition;
+    }
+  }
+
+  void _paintFloatingCaret(Canvas canvas, Offset effectiveOffset) {
+    assert(_textLayoutLastWidth == constraints.maxWidth);
+    final Paint paint = Paint()
+      ..color = _cursorColor;
+
+    final Rect caretPrototype = Rect.fromLTRB(_caretPrototype.left - _kFloatingCaretSizeOffset.dx,
+                                              _caretPrototype.top - _kFloatingCaretSizeOffset.dy,
+                                              _caretPrototype.right + _kFloatingCaretSizeOffset.dx,
+                                              _caretPrototype.bottom + _kFloatingCaretSizeOffset.dy);
+    final Rect caretRect = caretPrototype.shift(effectiveOffset);
+    const Radius floatingCursorRadius = Radius.circular(_kFloatingCaretRadius);
+    final RRect caretRRect = RRect.fromRectAndRadius(caretRect, floatingCursorRadius);
+    canvas.drawRRect(caretRRect, paint);
+  }
+
+  Offset _relativeOrigin = const Offset(0, 0);
```

Comment:
Couldn't figure out what these are from reading the names. Make the names more descriptive or add comments. Also be weary of making the names too generic if they're floating cursor specific.